presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -804,10 +804,17 @@ private PlanBuilder window(PlanBuilder subPlan, List<FunctionCall> windowFunctio
 
             List<Symbol> sourceSymbols = subPlan.getRoot().getOutputSymbols();
             ImmutableList.Builder<Symbol> orderBySymbols = ImmutableList.builder();
-            orderBySymbols.addAll(orderings.keySet());
+            orderBySymbols.addAll(sortOrders.keySet());
             Optional<OrderingScheme> orderingScheme = Optional.empty();
-            if (!orderings.isEmpty()) {
-                orderingScheme = Optional.of(new OrderingScheme(orderBySymbols.build(), orderings));
+            if (!sortOrders.isEmpty()) {
+                OrderingScheme ordering = new OrderingScheme(orderBySymbols.build(), sortOrders);
+
+                if (partitionBySymbols.build().isEmpty()) {
+                    // if ORDER BY but no PARTITION BY, insert global sort step before
```

Comment:
Why this is Planner and not a rule? Or enforcer in AddExchanges?

Reply:
From the commit message: 

> This could have been implemented as an optimizer rule. However,
the issue is that transforming a WindowNode(X) into a
WindowNode(Sort(X)) is, by itself, a suboptimal move, so it's
unnatural for a standalone optimization rule. It's only
due to the subsequent transformation of Sort(X) into a distributed
sort that the plan becomes more desirable than the original one.

> By implementing during the initial planning step, we start from
a less optimal plan that gets improved via rule transformation.

But I'm also trying with AddExchanges as an alternative. I fear it will get ugly. That class has got way too complicated.

Reply:
One advantage of doing it in the planner is that we can easily add a rule that converts:

```
- Window[row_number(ORDER BY a, b ,c)]
  - Sort[a, b, c]
    - X
```

into 

```
- RowNumber
  - Sort[a, b, c]
    - X
```

It'd be harder to do if the transformation happened inside AddExchanges. Eventually, instead of matching a Sort node we'd be matching on ordering trait, so node that produces that ordering would apply, not just SortNode.

Reply:
Another possible advantage is that a rule could improve cases like Window created _during_ planning, as an implementation of some other feature.
E.g. OFFSET (today it goes straight to RowNumberNode) or FETCH FIRST ... WITH TIES (perhaps a Window with `rank`).

> WindowNode(X) into a WindowNode(Sort(X)) is, by itself, a suboptimal move, so it's unnatural for a standalone optimization rule

Call it "pushdown Sort below Window" :) 
Of course you need to realize whether source is already sorted (property derivations / trait / visitor like `QueryCardinalityUtil`), otherwise this would loop.
Once you have this knowledge, the move is probably no longer suboptimal, is it?

Reply:
> However, the issue is that transforming a WindowNode(X) into a
WindowNode(Sort(X)) is, by itself, a suboptimal move, so it's
unnatural for a standalone optimization rule.

Could you elaborate a bit more? According to your comment when we insert `SortNode` during planning we start with suboptimal plan already. What's the difference from rule that pushes down sort? We can always check in such rule if distributed sort is enabled.

I think it's good heuristics to push down sort from window node. We loose a bit on memory but sort should happen faster.

Reply:
Sure. Going from `Window(X)` to `Window(Sort(X))` is better only if that is followed by a transformation to `Window(DistributedSort(X))`. Otherwise, `Window(Sort(X))` would require loading all the data in the sort operator, sorting it, outputting it, loading it again in the window operator, applying the function and emitting again. A `Window(X)` that sorts internally would skip the materialization and transfer between the operators, so it *should* be more efficient.

Therefore, the `Window(X)`->`Window(Sort(X))` moves the query plan to a less desirable local state (the analogy with a hill climbing-style optimizer would be moving down the hill towards a valley).

> We can always check in such rule if distributed sort is enabled.

Indeed. But it introduces more coupling between rules.